### PR TITLE
Automated tests (part 2) (connect #613)

### PR DESCRIPTION
**Pull request by MelEnt**
#### Before the PR (what is the issue or what needed to be done)
Tests previously required already installed and set up user. 
Also, this PR adds three new survey types (Date, Number and Options) which is built in the same manner as the previous survey (Freetext)

#### The solution
Proper setup and teardown procedures before and after tests, respectively, will hopefully solve the issue of missing prerequisites.

#### Screenshots (if appropriate)
NA

#### Reviewer Checklist
* [x] Connect the issue
* [ ] Test plan
* [ ] Copyright header
* [ ] Code formatting
* [ ] Documentation

### CODE REVIEW

File: app/src/androidTest/java/org/akvo/flow/activity/NumberSurveyTest.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+public class NumberSurveyTest {
+
+    @Rule
+    public ActivityTestRule<SurveyActivity> rule = new ActivityTestRule<>(SurveyActivity.class);
+    private SurveyInstaller installer;
+
+    @Before
+    public void init()
+    {
```
Comment (valllllll2000) [inline app/src/androidTest/java/org/akvo/flow/activity/NumberSurveyTest.java]: please fix the formatting of {}

File: app/src/androidTest/java/org/akvo/flow/testhelper/SurveyInstaller.java [hunk 2]
```diff
@@ -66,23 +88,25 @@ public Survey persistSurvey(String xml) throws IOException {
...
         adapter.addSurveyGroup(survey.getSurveyGroup());
         adapter.close();
-
         notifyNewSurvey(context);
-        return survey;
+    }
+
+    public void clearSurveys() {
+        for (File file : surveyFiles)
+        {
```
Comment (valllllll2000) [inline app/src/androidTest/java/org/akvo/flow/testhelper/SurveyInstaller.java]: please fix formatting

**Comment by valllllll2000** [comment]
@MelEnt Please apply the formatting (ctrl + alt + i and ctr + alt + l)  to all your files as the formatting is not correct and some lines are too long.

File: app/src/androidTest/java/org/akvo/flow/testhelper/SurveyRequisite.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        FlowApp.getApp().setUser(new User(1L, "User"));
+        Prefs prefs = new Prefs(context);
+        //To bypass the need for a setup (create user)
+        prefs.setBoolean(Prefs.KEY_SETUP, true);
+        //To bypass "Low Storage" prompt
+        prefs.setLong(Prefs.KEY_SPACE_AVAILABLE, 50L);
+    }
+
+    //Reset to stage before tests
+    public static void resetRequisites() {
```
Comment (valllllll2000) [inline app/src/androidTest/java/org/akvo/flow/testhelper/SurveyRequisite.java]: maybe clearRequisites would be a better name

File: app/src/androidTest/java/org/akvo/flow/activity/OptionsSurveyTest.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    @Test
+    public void canFillOptionsQuestion() throws IOException, InterruptedException {
+        fillOptionsQuestion(optionsurvey, 0);
+        onView(allOf(withClassName(endsWith("Button")), withText(R.string.submitbutton))).check(matches(isEnabled()));
+    }
+
+    private Survey fillOptionsQuestion(int surveyResId, int option) throws IOException {
+        Survey survey = getSurvey(surveyResId);
+
+        openDrawer();
```
Comment (valllllll2000) [inline app/src/androidTest/java/org/akvo/flow/activity/OptionsSurveyTest.java]: out of curiosity: why do you open the drawer here?

File: app/src/androidTest/java/org/akvo/flow/testhelper/SurveyRequisite.java (hunk 3 above)
Comment (MelEnt) [inline app/src/androidTest/java/org/akvo/flow/testhelper/SurveyRequisite.java]: Well, clear would imply they're gone, while reset tells us they're set to the stage before the tests

File: app/src/androidTest/java/org/akvo/flow/activity/OptionsSurveyTest.java (hunk 4 above)
Comment (MelEnt) [inline app/src/androidTest/java/org/akvo/flow/activity/OptionsSurveyTest.java]: Before selecting a survey we open the drawer. Even if it's already open, it doesn't hurt to double open it, just for the sake of assurance.

**Comment by valllllll2000** [comment]
@MelEnt @EnderCrypt I have been running all the tests again and the often fail. The reason for that is always the same:
```
android.support.test.espresso.NoMatchingViewException: No views in hierarchy found matching: with id: org.akvo.flow:id/new_datapoint
```
It seems that once you run one of the tests since we are in another activity (FormActivity) the new_datapoint menu button is no longer there so I would separate these tests in 2 parts: 
1) Test the survey activity, selecting survey from drawer and then pressing new_datapoint and make sure we are in FormActivity 
2) The actual tests for diferent question views which should be run using FormActivity started with the required survey.
Currently you seem to be testing the navigation itself multiple times but in my opinion this is not necessary. An alternative would be to maybe press the back button at the end of each test but I would prefer the first suggestion.

Also I have been running your tests with the following emulator config:
`emulator -use-system-libs -avd Pixel_XL_API_25 -no-audio -no-window`
Pixel_XL_API_25 is the name of the emulator
`-use-system-libs` is because my machine has some issues with emulator but you can probably omit that
and also I run the following settings before running the tests:
```
adb shell settings put global window_animation_scale 0 &
adb shell settings put global transition_animation_scale 0 &
adb shell settings put global animator_duration_scale 0 &
adb shell input keyevent 82
```
This is the way to run the tests in travis CI, that is why no window, once you make the fixes we will be able to run your tests using travis which will help us make sure new commits do not break the views.

**Comment by valllllll2000** [comment]
Will update the branch with fixes and create a new PR

**Pull request closed without merging**